### PR TITLE
search from autofill infrastructure

### LINKED ISSUE #421: Infrastructure to  search  entries when Autofilling
### Acceptance Criteria

When launching into the Lockbox search interface, the app switches appropriately from the Locked screen to a placeholder for the item list (or navigates directly to the placeholder when appropriate).

Related to #215

### LINKED ISSUE #465: Flesh out unit testing for (autofill) AuthPresenter
Largely due to some extraordinary circumstances, currently the `AuthPresenter` is untested.

The lack of tests @sashei pointed out when dsicussing how to get started testing #450.

### LINKED ISSUE #468: Fingerprint unlock dialog presentation on autofill
Currently the user is brought back to the Lockbox app to be presented with the Fingerprint unlock dialog. 

## Acceptance Criteria
When only authenticating to display entries in an autofill (i.e., "receiver" app) context, the fingerprint dialog will be displayed in that context rather than a Lockbox backdrop.

Depends on #421

### PR DESCRIPTION
**Pull request by sashei**
Fixes #421
Fixes #468 
Fixes #465 

## Testing and Review Notes

The expected behavior for the search infrastructure is as follows:
- when launching autofill with the app _locked_, tapping `Unlock Firefox Lockbox` prompt will bring you to the fingerprint dialog
    - if, after unlocking, there _are_ saved entries matching the autofill context, you are brought back to the autofill context with those entries suggested
    - if, after unlocking, there are no saved entries matching the autofill context, you are brought to the search context
- ~when launching autofill with the app _unlocked_, tapping `Search Firefox Lockbox` brings you to the search context~ (covered in #476 )
- selecting an entry in the search context will still show that entry as an autofill suggestion before you can select it (don't think we have any control there)
- tapping outside of the search box will cancel autofill

the search infrastructure will fill with a placeholder login for now. the search UI itself is getting built out at #471 

everything else implemented by autofill so far should behave as it did before (autofill field finding, domain parsing, etc)

The test data has google, facebook and twitter credentials in them. 

For demo and manual testing purposes, you can see the placeholder search box by asking for autofill in an app that is not in the test data. The Duolingo app or website works quite well. This works for locked and unlocked states.

You should be able to see unlocking and filling without the search with a google property.

### Notes:
1. it is possible to launch the autofill by long pressing on the autofillable input field.
2. Autofilling into Chrome does not launch our service.

## To Do

- [ ] double check the original issue to confirm it is fully satisfied
- [ ] add testing notes and screenshots in PR description to help guide reviewers
- [ ] add unit tests
- [ ] request the "UX" team perform a design review (if/when applicable)
- [ ] make sure CI builds are passing (e.g.: fix lint and other errors)
- [ ] check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI

### CODE REVIEW

**Comment by sashei** [comment]
@jhugman hand-off notes:
- the search display is working with the placeholder entry
- routing from the locked screen to the search screen is not working
- I just did the background styling, it was easier that way
- haven't tested tapping outside of the dialogs yet

File: app/src/main/java/mozilla/lockbox/presenter/AutofillRoutePresenter.kt [hunk 1]
```diff
@@ -58,14 +57,15 @@ class AutofillRoutePresenter(
...
-            .switchMap { dataStore.list }
-            .filter { it.isNotEmpty() }
-            .take(1)
-            .switchMap { responseBuilder.asyncFilter(pslSupport, Observable.just(it)) }
-            .filter { it.isNotEmpty() }
-            .subscribe(this::finishResponse)
+            .switchMap { responseBuilder.asyncFilter(pslSupport, dataStore.list.take(1)) }
+            .map { logins ->
+                if (logins.isNotEmpty()) {
+                    AutofillAction.CompleteMultiple(logins)
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/AutofillRoutePresenter.kt]: :chef_kiss:

File: app/src/main/java/mozilla/lockbox/presenter/AutofillLockedPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        dispatcher.dispatch(DataStoreAction.Unlock)
+    }
+
+    private fun unlockFallback() {
+        if (fingerprintStore.isKeyguardDeviceSecure) {
+            lockedView.unlockFallback()
+        } else {
+            unlock()
+        }
+    }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/presenter/AutofillLockedPresenter.kt]: #tech-debt: merge the `AutofillLockedPresenter` and `LockedPresenter`.

File: app/src/main/java/mozilla/lockbox/view/AutofillLockedFragment.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (resultCode == Activity.RESULT_OK) {
+            when (requestCode) {
+                LOCK_REQUEST_CODE -> _unlockConfirmed.onNext(true)
+            }
+        } else {
+            _unlockConfirmed.onNext(false)
+        }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/view/AutofillLockedFragment.kt]: What happens to `_unlockConfirmed` when a request code which isn't `LOCK_REQUEST_CODE` is received?

```kotlin
_unlockConfirmed.onNext(resultCode == Activity.RESULT_OK && requestCode == LOCK_REQUEST_CODE)
```

File: app/src/test/java/mozilla/lockbox/presenter/AutofillLockedPresenterTest.kt [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        Thread.sleep(600)
+        assertTrue(view.unlockFallbackCalled)
+    }
+
+    @Test
+    fun `when fingerprint auth is not available or enabled and there is other device security`() {
+        fingerprintStore.fingerprintAuthAvailableStub = false
+        fingerprintStore.keyguardDeviceSecureStub = true
+        settingStore.unlockWithFingerprintStub.onNext(false)
+        Thread.sleep(600)
```
Comment (jhugman) [inline app/src/test/java/mozilla/lockbox/presenter/AutofillLockedPresenterTest.kt]: 🤢 Gosh do we have to?

Comment (sashei) [reply app/src/test/java/mozilla/lockbox/presenter/AutofillLockedPresenterTest.kt]: I knowwww, I was having issues making the `delay(0)` work and got frustrated >.>

File: app/src/main/java/mozilla/lockbox/presenter/AutofillLockedPresenter.kt (hunk 2 above)
Comment (sashei) [reply app/src/main/java/mozilla/lockbox/presenter/AutofillLockedPresenter.kt]: hah I had the same thought during dinner last night.... they are sooo similar

**Pull request merged**